Add docs & tests to kubernetes exporter

zeenix opened the pull request:
Here we go. Apart from adding docs/tests, also included are a few small fixes/improvements.

Commit:
pkg-export-k8s: Add 2 simple unit tests to Manifest

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Add docs to Bind

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Add docs to Manifest

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Add docs to ManifestJson

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Use str::parse()

Instead of more explicit FromStr::from_str() to create a Topology from a
string.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: testcase for natural num validator

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Add docs to Cli

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Document export_for_cli_matches() function

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

krnowak commented on components/pkg-export-kubernetes/src/topology.rs:
```diff
@@ -18,6 +18,7 @@ use std::fmt;
 
 use error::Error;
 
+/// The relationship of an app with peers in the same ring
```
I think this is more of a relationship of an app in the same service group?

krnowak commented on components/pkg-export-kubernetes/src/topology.rs:
```diff
@@ -37,6 +59,16 @@ impl FromStr for Topology {
 }
 
 impl fmt::Display for Topology {
+    /// Basicially it converts a Topology into a string
```
I'd write "Converts a Topology into a string", without the "Basically" part (with a typo :P).

asymmetric replied:
Agreed. If this needs to be a documentation comment, it should sound like one.

krnowak commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -46,7 +46,7 @@ impl Manifest {
         let topology: Topology = FromStr::from_str(
             matches.value_of("TOPOLOGY").unwrap_or("standalone"),
-        ).unwrap_or(Topology::Standalone);
+        ).unwrap_or(Default::default());
```
You probably could have `let topology = FromStr(…).unwrap_or(Topology::default())`, no? (Without the `: Topology` part after the variable name).

asymmetric replied:
I think something along the lines of `Topology::from_str` should work.

krnowak commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -105,3 +105,62 @@ impl Manifest {
+            habitat_name: "hi".to_owned(),
+            image: "core/hi".to_owned(),
+            count: 3,
+            service_topology: Default::default(),
```
Wouldn't `Topology::default()` be clearer?

zeenix replied:
from the docs, it seems Default::default() is prefered as type is in the declaration anyway.

krnowak commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -105,3 +105,62 @@ impl Manifest {
+            count: 3,
+            service_topology: Default::default(),
+            service_group: Some("group1".to_owned()),
+            config_secret_name: Some("alphabravocharlie".to_owned()),
```
Too much flying. ;)

lilic replied:
:woman_facepalming: 😆

zeenix replied:
haha, @krnowak unfortunately not even enough flying. :( @LiliC i didn't quite get the emoji combo :)

krnowak commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -28,20 +28,39 @@ use topology::Topology;
     pub habitat_name: String,
+    /// The docker image.
     pub image: String,
+    /// The amount of Services that should start in the Habitat ring.
```
Maybe instead of "in the Habitat ring", write "in the service group"?

lilic replied:
Yes, this doesn't have anything to do with the ring. This has more to do with the number of instances you want to start.

zeenix replied:
isn't the "ring" a similar concept to "cluster" of k8s?

asymmetric replied:
It's better not to make parallels, as the worlds are really quite different :)

krnowak commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -28,20 +28,39 @@ use topology::Topology;
     pub image: String,
+    /// The amount of Services that should start in the Habitat ring.
     pub count: u64,
+    /// The relationship of an app with peers in the same ring.
```
Same here.

lilic replied:
App is not the phrase they use... think "service" would be a better phrasing.

krnowak commented on components/pkg-export-kubernetes/src/bind.rs:
```diff
@@ -27,6 +28,32 @@ pub struct Bind {
+    /// assert_eq!(b.service, "s");
+    /// assert_eq!(b.group, "g");
+    ///
+    /// for s in ["ns:g", "nsg", "nsg", ":nsg:"].iter() {
```
You have repeated "nsg" twice here.

krnowak commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -63,9 +63,11 @@ impl Manifest {
+            .value_of("TOPOLOGY")
+            .unwrap_or("standalone")
+            .parse()
+            .unwrap_or(Default::default());
```
How about using `Topology::default()`? Would it spare you from explicitly specifying the Topology type next to the variable name?

zeenix replied:
I'm not sure. Even the usage of topology later might be enough to tell rustc which type to pick. I'll check..

asymmetric replied:
I don't see the point of this change, since now we are using the default two times, which is ugly. I think it looked better before, with maybe `String::from` instread of `FromStr::from_str`.

zeenix replied:
Are we? I only see 1 call to Default::default() here. If you mean the both components of the call then it was the same previously as well. :) The change here is not the Default::default but rather Fromstr::from_str() -> parse(), which is what I was explicitly asked to do in the Helm exporter PR by cm and is the general convention in Rust world afaik. If you had reviewed individual commits instead, it'd have been very obvious. :)

asymmetric replied:
I did review individual commits. The point about two defaults is the `unwrap_or("standalone")` and `unwrap_or(Default:::default())`.

zeenix replied:
The actual value of the default is quite unimportant and redundant IMO here. "Just use the default topology" is what we are saying here.

krnowak commented on components/pkg-export-kubernetes/src/cli.rs:
```diff
@@ -162,3 +162,18 @@ fn valid_natural_number(val: String) -> result::Result<(), String> {
+
+    #[test]
+    fn test_valid_natural_number() {
+        valid_natural_number("99".to_owned()).unwrap();
```
Are hexadecimal strings also accepted?

zeenix replied:
No. I know what to do. :)

lilic commented on components/pkg-export-kubernetes/src/cli.rs:
```diff
@@ -76,6 +87,7 @@ impl<'a, 'b> Cli<'a, 'b> {
         }
     }
 
+    /// Add Habitat (operator) runtime arguments to the CLI.
```
Bit confused by this comment. Can you explain what this does?

zeenix replied:
We're adding commandline options/arguments in each of these methods. This one is adding the ones that provide values for the k8s operator API. Not sure that makes it clear though?

lilic replied:
But it has nothing to do with the operator itself.

zeenix replied:
how come? They are the operator API, aren't they?

lilic replied:
Sort of, we pass them to Habitat service itself.

zeenix replied:
That's why I wrote operator in brackets. :)

krnowak reviewed (changes requested):
Looks mostly fine, some nits.

lilic commented on components/pkg-export-kubernetes/src/lib.rs:
```diff
@@ -50,6 +50,9 @@ pub use topology::Topology;
 
+/// Convenient do-it-all function. You give it the CLI arguments from the user and it generates the
+/// Kubernetes manifest. If user passed an `--output` argument with a value that is not "`-`", the
+/// manifest is written to the provided file; otherwise, it is written on the standard output.
```
Isn't it "written to the standard output"?

asymmetric replied:
This sounds more like an inline comments than a documentation comment, but if that's your style.. :)

lilic commented on components/pkg-export-kubernetes/src/bind.rs:
```diff
@@ -17,6 +17,7 @@ use std::str::FromStr;
 
 use error::Error;
 
+/// A Habitat bind inside a Kubernetes cluster
```
"A Habitat service binding... "

lilic commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -28,25 +28,46 @@ use topology::Topology;
     pub count: u64,
+    /// The relationship of an app with peers in the same ring.
     pub service_topology: Topology,
+    /// The logical group of services in the Habitat ring.
```
"The logical grouping of services..."

zeenix replied:
grouping is the verb, the result of which is a "group". We are just specifying the group.

asymmetric replied:
I agree with Lili on this one, it sounds weird like this. 🤷‍♂️

zeenix replied:
weird maybe but i think it's correct, like I explained above.

lilic commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -28,25 +28,46 @@ use topology::Topology;
     pub service_topology: Topology,
+    /// The logical group of services in the Habitat ring.
     pub service_group: Option<String>,
+    /// The name of the Kubernetes secret containing the config file - user.toml - that the user
```
"The name of the Kubernetes secret containing the config file, which will serve as initial configuration...."

asymmetric replied:
I would remove mentions of the file name, as that could change across the project without too much effort. I added a [constant](
https://github.com/kinvolk/habitat/tree/23a68eef345a431b385317291fb1fdcfe207b14b/components/core/src/fs.rs#L47) for this reason in my branch.

zeenix replied:
@asymmetric you mean "- user.toml -" bit above should be removed?

lilic commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -105,3 +127,62 @@ impl Manifest {
+        let mut m = Manifest {
+            metadata_name: "hi-f84iq".to_owned(),
+            habitat_name: "hi".to_owned(),
+            image: "core/hi".to_owned(),
```
That's not how you specify an image name. It would be `remote:service_name:tag_name` so `core/hi:latest` for example.

zeenix replied:
Oh but that's not how k8s exporter has been doing it.

lilic replied:
Well then it maybe should. ;)

asymmetric replied:
I guess there's something around `:latest` being implicit if not present.

zeenix replied:
yeah and that would explain why it has been working?

lilic commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -105,3 +127,62 @@ impl Manifest {
+        let mut m = Manifest {
+            metadata_name: "hi-f84iq".to_owned(),
+            habitat_name: "hi".to_owned(),
+            image: "core/hi".to_owned(),
```
Same as above about the image name...

zeenix replied:
This is the name of the docker image.

lilic replied:
But there is no tag after the image name?

zeenix replied:
no, currently not.

lilic commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -105,3 +127,62 @@ impl Manifest {
+    fn test_manifest_generation_binds() {
+        let mut m = Manifest {
+            metadata_name: "hi-f84iq".to_owned(),
+            habitat_name: "hi".to_owned(),
```
I am a bit confused what this `habitat_name` is?

zeenix replied:
in the manifest template:

```
metadata:
  ## Name of the Habitat service.
  name: {{metadata_name}}                                                                                                              
  labels:
    habitat-name: {{habitat_name}}
```

IIRC it used to be just one name here but then Michal split into two names and added a random string to the former so that there can be multiple instances of a service.

lilic commented on components/pkg-export-kubernetes/src/manifestjson.rs:
```diff
@@ -24,12 +24,24 @@ use manifest::Manifest;
 const BINDFILE: &'static str = include_str!("../defaults/KubernetesBind.hbs");
 
+/// Represents the [`Manifest`] in JSON format. This is an intermediate type that can be converted
+/// to the final manifest YAML file content, ready for consumption by your Kubernetes cluster.
```
"....by a Kubernetes cluster."

zeenix replied:
what's wrong with making it a bit personal? :)

asymmetric replied:
LOL, usually it's @LiliC who writes personalized documentation :P

lilic replied:
Well I learned from your suggestions not to. 😉

zeenix replied:
oh?

lilic commented on components/pkg-export-kubernetes/src/manifestjson.rs:
```diff
@@ -60,6 +72,12 @@ impl ManifestJson {
     }
 
+    /// Convert into a string. The returned string is the final manifest YAML file content, ready
+    /// for consumption by your Kubernetes cluster.
```
""....by a Kubernetes cluster.""

lilic commented on components/pkg-export-kubernetes/src/topology.rs:
```diff
@@ -18,7 +18,8 @@ use std::fmt;
 use error::Error;
 
-#[derive(Debug, Clone)]
+/// The relationship of an app with peers in the same ring
```
In some places you do full sentences, in others you don't. Not sure what the rule for this in Rust?

lilic replied:
"The relationship of a service with peers in the same ring"

asymmetric replied:
Don't think there's a rule as in Go, but I create full sentences with periods. Don't think the Chef folks do it all the time though.

Considering this is a doc-comment, it should probably be a full sentence with a period.. or do they get added automatically?

zeenix replied:
I think the doc tool just requires you to be consistent with it and if you don't use '.', it adds them for you. Anyway, I should be consistent here.

lilic commented on components/pkg-export-kubernetes/src/topology.rs:
```diff
@@ -37,9 +59,25 @@ impl FromStr for Topology {
     }
 }
+
+impl Default for Topology {
```
Is this the same default as in Habitat itself? Can't we get it from the codebase rather then redefining it?

asymmetric replied:
Good point, it's defined [here](
https://github.com/kinvolk/habitat/tree/23a68eef345a431b385317291fb1fdcfe207b14b/components/sup/src/manager/service/mod.rs#L893) already.

zeenix replied:
Good catch. We have the same implementation of Topology in the supervisor crate. The question is if we want to depend on that crate here?

lilic commented on components/pkg-export-kubernetes/tests/KubernetesManifestTest.yaml:
```diff
@@ -0,0 +1,38 @@
+#  name: user-toml-secret
+#type: Opaque
+#data:
+#  # Each configuration item needs to be encoded in base64.
```
One space too many. So it should be `## Each ...`

zeenix replied:
Mos of this is just a copy&paste of the template we've been using.

zeenix replied:
Actually that's a comment inside a commented-out block but i'll change it that there is no commented out stuff anyway.

lilic commented on components/pkg-export-kubernetes/tests/KubernetesManifestTest.yaml:
```diff
@@ -0,0 +1,38 @@
+  ## Name of the Habitat service.
+  name: hi-f84iq
+  labels:
+    habitat-name: hi
```
I would prefer to use a real life service example, it makes more sense. So for example nginx or mysql, or something similar to that. WDYT?

zeenix replied:
I don't see the point. This is just a unit test part. I hope nobody uses it for anything else anyway. :)

lilic replied:
Just feel like it's more readable and understandable.

asymmetric replied:
I think random strings are fine for tests.

zeenix replied:
true. I'll change that part too then.

lilic commented on components/pkg-export-kubernetes/tests/KubernetesManifestTest.yaml:
```diff
@@ -0,0 +1,38 @@
+    habitat-name: hi
+spec:
+  ## Name of the Habitat service package exported as a Docker image.
+  image: core/hi
```
As above, you would most likely use the tag as well when creating a manifest file. So for example: `core/hi:latest`.

lilic commented on components/pkg-export-kubernetes/tests/KubernetesManifestTest.yaml:
```diff
@@ -0,0 +1,38 @@
+spec:
+  ## Name of the Habitat service package exported as a Docker image.
+  image: core/hi
+  ## Number of desired instances.
```
See here you say it's an instance, so you should do the same above.

zeenix replied:
Just different wording two different people came up with, but sure. :)

lilic commented on components/pkg-export-kubernetes/tests/KubernetesManifestTest.yaml:
```diff
@@ -0,0 +1,38 @@
+    ## Habitat service group name, a logical grouping of services
+    ## with the same package.
+    group: group1
+    ## Name of the configuration secret. Edit the Kubernetes Secret
```
"Name of the secret containing the initial configuration toml file."

lilic commented on components/pkg-export-kubernetes/tests/KubernetesManifestTest.yaml:
```diff
@@ -0,0 +1,38 @@
+---
```
A bit confused by this test. You have the secret commented out, but you end up using it on 35 of this file. I would get rid of all the mention of secret and the commented out code. Otherwise this manifest file would error out in a k8s cluster as it would expect a secret to be present.

zeenix replied:
Where is the secret commented out in?

lilic replied:
Right below this comment, the secret creation is commented out.

zeenix replied:
Oh! I didn't realize we are creating a secret in this commented out code. This has been the part of the manifest that was generated already by the first k8s exporter @krnowak wrote and I just kept this as is. I was supposed to ask about it but then I forgot. :(

Anyway, so this part is supposed to create the file that configSecretName later specifies?

lilic replied:
Yes, exactly.

I would suggest the following test scenarios if I understand correctly:
1. Basic manifest file with no configSecretName and the secret commented out.
2. Bind manifest file.
3. Manifest file that has everything uncommented.

zeenix replied:
Alright, i need to fix the exporter code then, Although we don't want commented out stuff though. If secret is not specified, it's just not there in the manifest.

lilic replied:
Yep, agreed. 👍  I thought that's how it is already...

lilic commented on components/pkg-export-kubernetes/tests/KubernetesManifestTestBinds.yaml:
```diff
@@ -0,0 +1,45 @@
+    group: group1
+    ## Name of the configuration secret. Edit the Kubernetes Secret
+    ## at the top.
+    configSecretName: alphabravocharlie
```
Again as above, do we need to have the secret name, this would not work if the secret is not created.

zeenix replied:
we don't need it, no but this is just a sample configuration for testing (i-e if we specify the secret to the exporter, does it correctly end up in the generated manifest or not. This is not meant for deploying in the cluster.

lilic replied:
I mean, this is not how we want the file to end up looking. If I understand correctly this is the resulting file, right?
Then it should have the correctly generated manifest file. If the secret is commented out but the `configSecretName` is there, that should not be a valid file.

asymmetric commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -105,3 +105,62 @@ impl Manifest {
+
+        let out = String::from_utf8(o).unwrap();
+
+        assert!(out == expected);
```
[`assert_eq!`](https://doc.rust-lang.org/std/macro.assert_eq.html) would make more sense here IMO.

asymmetric commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -105,3 +105,62 @@ impl Manifest {
+            image: "core/hi".to_owned(),
+            count: 3,
+            service_topology: Default::default(),
+            service_group: Some("group1".to_owned()),
```
I tend to use `String::from` in these cases, unless I already have a `&str`. 🤷‍♂️

zeenix replied:
hmm.. I would like to know what the convention is actually. I have a feeling it's String:from()

zeenix replied:
I was informed that it's all personal preferences. :)

asymmetric commented on components/pkg-export-kubernetes/tests/KubernetesManifestTest.yaml:
```diff
@@ -0,0 +1,38 @@
+  image: core/hi
+  ## Number of desired instances.
+  count: 3
+  ## An object containing parameters that effects how the Habitat service
```
affects

asymmetric commented on components/pkg-export-kubernetes/tests/KubernetesManifestTest.yaml:
```diff
@@ -0,0 +1,38 @@
+apiVersion: habitat.sh/v1
+kind: Habitat
+metadata:
+  ## Name of the Habitat service.
```
Why use two `#`?

zeenix replied:
I just checked from git blame, it blames @krnowak for this. :) i-e Two # were used in the first manifest template string used in 8598b4137d0ec14def123acb5f77ba9ee8b39297.

asymmetric replied:
OK, should be removed then I guess?

asymmetric commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -28,20 +28,39 @@ use topology::Topology;
 pub struct Manifest {
+    /// The service name in the Kubernetes cluster.
     pub metadata_name: String,
+    /// The service name in the Habitat ring.
```
I think the difference between these two name fields should be explained better. I don't understand it for example.

zeenix replied:
Sure, that way maybe I'll get it in my head too. :)

asymmetric commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -28,20 +28,39 @@ use topology::Topology;
     pub metadata_name: String,
+    /// The service name in the Habitat ring.
     pub habitat_name: String,
+    /// The docker image.
```
Should there be a separate field for the tag?

zeenix replied:
Not sure I follow. What tag? This is the same as the image name: origin/service

asymmetric replied:
The docker tag. It could in theory be an additional `String` field. Or we could just expect it to be present (or not) in this field. Not sure about the pros/cons.

zeenix replied:
ah yes. it should be a String but we'll address that later.

asymmetric commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -28,20 +28,39 @@ use topology::Topology;
+
+    /// Any binds, as [`Bind`] instances.
+    ///
+    /// [`Bind`]: ../bind/struct.Bind.html
```
Have you tested that these relative links work? Just asking cause it took me a while to find out what the correct relative path was.

zeenix replied:
I did. I hate it that you have to point to specific html like this in the source comments. At least types from the same crate should not need to be linked.

asymmetric commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -28,20 +28,39 @@ use topology::Topology;
 
 impl Manifest {
+    ///
+    /// Create a Manifest instance from commandline arguments passed as [`clap::ArgMatches`].
```
commandline is not a word: command-line.

asymmetric commented on components/pkg-export-kubernetes/src/manifestjson.rs:
```diff
@@ -24,12 +24,24 @@ use manifest::Manifest;
+/// Represents the [`Manifest`] in JSON format. This is an intermediate type that can be converted
+/// to the final manifest YAML file content, ready for consumption by your Kubernetes cluster.
+///
+/// The reason for the existance of this intermediate type is to allow users of this crate to be
```
existence

asymmetric commented on components/pkg-export-kubernetes/src/manifestjson.rs:
```diff
@@ -24,12 +24,24 @@ use manifest::Manifest;
+/// to the final manifest YAML file content, ready for consumption by your Kubernetes cluster.
+///
+/// The reason for the existance of this intermediate type is to allow users of this crate to be
+/// able to modify the JSON before converting it into the final manifest string.
```
Nit: here you say "converted into"  but above "converted to". Intentional?

zeenix replied:
not intentional but I think both are correct but I won't argue against consistency. :)

asymmetric commented on components/pkg-export-kubernetes/src/manifestjson.rs:
```diff
@@ -60,6 +72,12 @@ impl ManifestJson {
+    ///
+    /// # Errors
+    ///
+    /// * rendering from the template fails (unlikely)
```
Unlikely is not very informative. Maybe specify when exactly it can happen or remove it altogether?

zeenix replied:
It tells the reader that they shouldn't be too scared of just unwrap()ing the result. Since the template is provided by us (from a separate file in the repository), the only reason it can possibly fail is a programmer error (messing up the template) or the cmdline options not properly checked (sill a programmer error). Not sure how to write that here.

asymmetric replied:
"Rendering from the template fails. This can only happen if the template has a syntax error or the command-line arguments are inconsistent" - or something like that.

zeenix replied:
Actually since it's likely going to be caused by a programming error, I wonder if this method should should just return a String and use expect() for errors. That way I can just impl Into<String> and get rid of the FIXME too.

asymmetric commented on components/pkg-export-kubernetes/src/bind.rs:
```diff
@@ -27,6 +28,32 @@ pub struct Bind {
+    ///
+    /// for s in ["ns:g", "nsg", "nsg", ":nsg:"].iter() {
+    ///     if let Err(e) = s.parse::<Bind>() {
+    ///         match e {
```
Wouldn't it be easier to use assert instead of this match? Not sure how exactly this would look, but in pseudocode:

    assert_eq!(s.type(), MyError)

zeenix replied:
that was what I wanted to do but the String inside the InvalidBind doesn't allow simple equality test. I asked on the IRC channel and this was the best way I was told is possible.

asymmetric commented on components/pkg-export-kubernetes/src/cli.rs:
```diff
@@ -162,3 +162,18 @@ fn valid_natural_number(val: String) -> result::Result<(), String> {
+    fn test_valid_natural_number() {
+        valid_natural_number("99".to_owned()).unwrap();
+
+        for &s in ["x", "", "#####"].iter() {
```
For my understanding, what's the point of `&s` here, instead of `s`?

zeenix replied:
So not to take ownership and borrow as string slice. IIRC it didn't build without the "&".

asymmetric commented on components/pkg-export-kubernetes/src/cli.rs:
```diff
@@ -27,12 +31,19 @@ where
         Cli { app: app }
     }
 
+    /// Convenient method to add all known arguments to the CLI.
```
Convenience method? Otherwise you're singing your own praises :)

zeenix replied:
It means the same thing. I know "convenience" is used typically but an adjective sounds more correct to me.

asymmetric replied:
This is getting OT but how can convenient and convenience mean the same thing?

zeenix replied:
One is derived from the other (don't know which way around), one is an adjective and the other is a noun. Which is why it sounds wrong to me to use the noun form as an adjective.

asymmetric reviewed (changes requested):
Good job all around, just some comments (about comments mostly)

Commit:
pkg-export-k8s: Require base64 crate

In a following patch, we'll use this to encode configuration file content
into base64.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: user-config-secret -> config

The handling of the configuration file was all wrong: The part of the
manifest that is supposed to create the configuration file, was always
commented out and instead of taking the configuration file from the user,
we were expecting them to provide the name for the file. This patch fixes
this issue.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
sup: Add a doc comment to Topology enum

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
sup: Slightly more thorough Topology tests

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
sup: Fix formatting to satisfy rustfmt

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Ditch custom Topology type

This type with the same API and tests already exists in the habitat_sup
crate so let's just make use of that and ditch our version.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
sup: Fix some formatting issues

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
sup: Make use of the result

Not only this is the right thing to do, it also get rid of compiler
warnings about this.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
sup: Drop a redundant 'mut'

One less compiler warning.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
sup: Drop a redundant 'using' statement

One less compiler warning.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Fix a minor grammatical mistake

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

zeenix commented:
I just forced-push and I believe I've addressed all comments other than the naming of service and image. I'll address that next but would there be any objection to addressing that in a branch on top of this?

lilic commented on .bldr.toml:
```diff
@@ -176,6 +176,7 @@ paths = [
   "components/common/*",
   "components/hab/*",
   "components/pkg-export-docker/*"
```
Is a comma not needed here?

asymmetric replied:
In case people wanna install a syntax checker, [this](https://github.com/vmchale/tomlcheck) seems to be the only one.

zeenix replied:
Indeed it is. Thanks for catching this. I don't have any means to test my changes to the CI (or i don't know them).

zeenix replied:
but yeah, I can at least test the toml syntax with that tool. :)

zeenix replied:
That tool sucks btw. if you make a mistake in toml, it doesn't tell you where exactly but still useful to verify that you've fixed everything, once you figure out on your own what's wrong. :)

lilic commented on .bldr.toml:
```diff
@@ -186,6 +187,7 @@ paths = [
   "components/hab/*",
   "components/pkg-export-docker/*"
```
Same question here for the comma?

lilic commented on components/pkg-export-kubernetes/defaults/KubernetesManifest.hbs:
```diff
@@ -1,15 +1,16 @@
+  name: user-toml-secret
+type: Opaque
+data:
+  ## Each configuration item needs to be encoded in base64.
```
Should we add a warning that this is just encoding and these files should not be shared if they contain sensitive configuration, or is it obvious?

zeenix replied:
are people really expected to modify these generated files?

zeenix replied:
but I would hope that if someone knows what base64 is and can encode things into that, they know that it's not encrypted.

lilic replied:
Most likely they will need to modify them to suit their needs. Yes, one would hope. :)

lilic commented on components/pkg-export-kubernetes/src/cli.rs:
```diff
@@ -107,34 +119,32 @@ impl<'a, 'b> Cli<'a, 'b> {
+                        .long("config")
+                        .short("n")
+                        .help(
+                            "The path to habitat configuration file in user.toml format. Habitat \
```
Habitat is uppercase.

lilic commented on components/pkg-export-kubernetes/src/cli.rs:
```diff
@@ -107,34 +119,32 @@ impl<'a, 'b> Cli<'a, 'b> {
+                        .short("n")
+                        .help(
+                            "The path to habitat configuration file in user.toml format. Habitat \
+                            will use it for initial configuration of the service",
```
of the service running inside of Kubernetes.

zeenix replied:
well that's clear from the context I'd say but I'll add it since you prefer. :)

lilic commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -105,3 +136,62 @@ impl Manifest {
+        let mut m = Manifest {
+            metadata_name: "nginx-f84iq".to_owned(),
+            habitat_name: "nginx".to_owned(),
+            image: "core/nginx".to_owned(),
```
We still don't have the tag?

zeenix replied:
but the "latest" tag is implicit here. I thought I'll only add the tag when we have version (and maybe release) in the service name and I think we agreed to do that in a separate PR after this.

lilic replied:
Yes, separate PR is fine for this. I still like to do `latest` by default rather then omitting the tag.

lilic commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -105,3 +136,62 @@ impl Manifest {
+        let mut m = Manifest {
+            metadata_name: "nginx-f84iq".to_owned(),
+            habitat_name: "nginx".to_owned(),
+            image: "core/nginx".to_owned(),
```
Same question here about the missing image tag?

lilic reviewed (changes requested):
Looking better, just a few questions. 🙂

Commit:
Add missing commas in .bldr.toml

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Explicit ":latest" tag on docker image

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

zeenix commented:
I just made another forced push. I also tested the changes and except for the usual issues, everything is as it always was. Please have another quick look and then I submit upstream.

zeenix commented:
Closing this now. Further reviews upstream..
